Code/GraphMol/ROMol.cpp

```diff
@@ -194,115 +193,119 @@ unsigned int ROMol::getNumAtoms(bool onlyExplicit) const {
-  MolGraph::vertex_descriptor vd = boost::vertex(idx, d_graph);
-  const Atom *res = d_graph[vd];
+  auto vd = boost::vertex(idx, d_graph);
+  const auto res = d_graph[vd];
 
   POSTCONDITION(res, "");
   return res;
 }
 
 // returns the first inserted atom with the given bookmark
 Atom *ROMol::getAtomWithBookmark(int mark) {
-  PRECONDITION(d_atomBookmarks.count(mark) != 0, "atom bookmark not found");
-  PRECONDITION(d_atomBookmarks[mark].begin() != d_atomBookmarks[mark].end(),
+  auto lu = d_atomBookmarks.find(mark);
+  PRECONDITION((lu != d_atomBookmarks.end() && !lu->second.empty()),
                "atom bookmark not found");
-  return *(d_atomBookmarks[mark].begin());
+  return lu->second.front();
 };
 
 // returns all atoms with the given bookmark
 ROMol::ATOM_PTR_LIST &ROMol::getAllAtomsWithBookmark(int mark) {
-  PRECONDITION(d_atomBookmarks.count(mark) != 0, "atom bookmark not found");
-  return d_atomBookmarks[mark];
+  auto lu = d_atomBookmarks.find(mark);
+  PRECONDITION(lu != d_atomBookmarks.end(), "atom bookmark not found");
+  return lu->second;
 };
 
 // returns the unique atom with the given bookmark
 Atom *ROMol::getUniqueAtomWithBookmark(int mark) {
-  PRECONDITION(d_atomBookmarks.count(mark) == 1,
-               "multiple atoms with same bookmark");
-  return getAtomWithBookmark(mark);
+  auto lu = d_atomBookmarks.find(mark);
+  PRECONDITION((lu != d_atomBookmarks.end()), "bookmark not found");
+  return lu->second.front();
 }
 
 // returns the first inserted bond with the given bookmark
 Bond *ROMol::getBondWithBookmark(int mark) {
-  PRECONDITION(d_bondBookmarks.count(mark) != 0, "bond bookmark not found");
-  PRECONDITION(d_bondBookmarks[mark].begin() != d_bondBookmarks[mark].end(),
+  auto lu = d_bondBookmarks.find(mark);
+  PRECONDITION((lu != d_bondBookmarks.end() && !lu->second.empty()),
                "bond bookmark not found");
-  return *(d_bondBookmarks[mark].begin());
+  return lu->second.front();
 };
 
 // returns all bonds with the given bookmark
 ROMol::BOND_PTR_LIST &ROMol::getAllBondsWithBookmark(int mark) {
-  PRECONDITION(d_bondBookmarks.count(mark) != 0, "bond bookmark not found");
-  return d_bondBookmarks[mark];
+  auto lu = d_bondBookmarks.find(mark);
+  PRECONDITION(lu != d_bondBookmarks.end(), "bond bookmark not found");
+  return lu->second;
 };
 
 // returns the unique bond with the given bookmark
 Bond *ROMol::getUniqueBondWithBookmark(int mark) {
-  PRECONDITION(d_bondBookmarks.count(mark) == 1,
-               "multiple bonds with same bookmark");
-  return getBondWithBookmark(mark);
+  auto lu = d_bondBookmarks.find(mark);
+  PRECONDITION((lu != d_bondBookmarks.end()), "bookmark not found");
+  return lu->second.front();
 }
 
 void ROMol::clearAtomBookmark(const int mark) { d_atomBookmarks.erase(mark); }
 
 void ROMol::clearAtomBookmark(int mark, const Atom *atom) {
-  if (d_atomBookmarks.count(mark) != 0) {
-    ATOM_PTR_LIST *entry = &d_atomBookmarks[mark];
+  PRECONDITION(atom, "no atom");
```

Comment:
We should probably own this atom as well.

Reply:
I don't believe that there is (or needs to be) any such requirement.